Client/Frontend/Widgets/AutocompleteTextField.swift

```diff
@@ -0,0 +1,139 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+
+/// Delegate for the text field events. Since AutocompleteTextField owns the UITextFieldDelegate,
+/// callers must use this instead.
+protocol AutocompleteTextFieldDelegate: class {
+    func autocompleteTextField(autocompleteTextField: AutocompleteTextField, didTextChange text: String)
+    func autocompleteTextFieldShouldReturn(autocompleteTextField: AutocompleteTextField) -> Bool
+    func autocompleteTextFieldShouldClear(autocompleteTextField: AutocompleteTextField) -> Bool
+    func autocompleteTextFieldDidBeginEditing(autocompleteTextField: AutocompleteTextField)
+    func autocompleteTextFieldDidEndEditing(autocompleteTextField: AutocompleteTextField)
+}
+
+/// Data source for the text field autocompletion.
+protocol AutocompleteTextFieldDataSource: class {
+    func autocompleteTextField(autocompleteTextField: AutocompleteTextField, suggestionForText text: String, complete: String? -> Void)
+}
+
+private struct AutocompleteTextFieldUX {
+    static let HighlightColor = UIColor(rgb: 0xccdded)
+}
+
+class AutocompleteTextField: UITextField, UITextFieldDelegate {
+    weak var autocompleteDelegate: AutocompleteTextFieldDelegate?
+    weak var autocompleteDataSource: AutocompleteTextFieldDataSource?
+
+    private var autocompleting = false
+
+    override init() {
+        super.init()
+    }
+
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+        super.delegate = self
+    }
+
+    required init(coder aDecoder: NSCoder) {
+        super.init(coder: aDecoder)
+        super.delegate = self
+    }
+
+    private var enteredText: NSString = "" {
+        didSet {
+            autocompleteDelegate?.autocompleteTextField(self, didTextChange: enteredText)
+        }
+    }
+
+    func textField(textField: UITextField, shouldChangeCharactersInRange range: NSRange, replacementString string: String) -> Bool {
```

Comment:
I wonder if you could just use a didSet-method on the text attribute and do basically the same thing here (then you don't have to stomp on this delegate).